libr/anal/esil_dfg.c

```diff
@@ -26,14 +30,373 @@ void _dfg_node_free (RAnalEsilDFGNode *free_me) {
+#endif
+		return 0;
+	}
+
+/*
+    =incoming=
+===in===
+
+shrink in
+
+    =incoming=
+=in=
+*/
+
+	if (rv_in->to <= rv_incoming->to) {
+#if DFG_DEBUG
+		eprintf ("del_alloc_cmp: case 7 - return 1\n");
+		eprintf ("del_alloc_cmp: adjusting in:\n\t[%d-%d] => ", rv_in->from, rv_in->to);
+#endif
+		rv_in->to = rv_incoming->from - 1;
+#if DFG_DEBUG
+		eprintf ("[%d-%d]\n", rv_in->from, rv_in->to);
+#endif
+		return 1;
+	}
+
+/*
+   =incoming=
+         ===in===
+
+up-shrink in
+
+   =incoming=
+             =in=
+
+*/
+
+#if DFG_DEBUG
+	eprintf ("del_alloc_cmp: case 8 - return -1\n");
+	eprintf ("del_alloc_cmp: adjusting in:\n\t[%d-%d] => ", rv_in->from, rv_in->to);
+#endif
+	rv_in->from = rv_incoming->to + 1;
+#if DFG_DEBUG
+	eprintf ("[%d-%d]\n", rv_in->from, rv_in->to);
+#endif
+	return -1;
 }
 
-static void* _edf_reg_get(RAnalEsilDFG *edf, const char *reg) {
-	return edf ? sdb_ptr_get (edf->latest_nodes, reg, 0) : NULL;
+static int _rv_ins_cmp (void *incoming, void *in, void *user) {
+	EsilDFGRegVar *rv_incoming = (EsilDFGRegVar *)incoming;
+	EsilDFGRegVar *rv_in = (EsilDFGRegVar *)in;
+	return rv_incoming->from - rv_in->from;
+}
+
+static bool _edf_reg_set (RAnalEsilDFG *dfg, const char *reg, RGraphNode *node) {
+	if (!dfg || !sdb_num_exists (dfg->regs, reg)) {
+		return false;
+	}
+	EsilDFGRegVar *rv = R_NEW0 (EsilDFGRegVar);
+	if (!rv) {
+		return false;
+	}
+	const ut64 v = sdb_num_get (dfg->regs, reg, NULL);
+	rv->from = (v & (UT64_MAX ^ UT32_MAX)) >> 32;
+	rv->to = v & UT32_MAX;
+	r_queue_enqueue (dfg->todo, rv);
+	while (!r_queue_is_empty (dfg->todo)) {
+	// rbtree api does sadly not allow deleting multiple items at once :(
+		rv = r_queue_dequeue (dfg->todo);
+		r_rbtree_cont_delete (dfg->reg_vars, rv, _rv_del_alloc_cmp, dfg);
+		if (dfg->insert) {
+			r_rbtree_cont_insert (dfg->reg_vars, dfg->insert, _rv_ins_cmp, NULL);
+			dfg->insert = NULL;
+		}
+		free (rv);
+	}
+	rv = R_NEW0 (EsilDFGRegVar);
+	rv->from = (v & (UT64_MAX ^ UT32_MAX)) >> 32;
+	rv->to = v & UT32_MAX;
+	rv->node = node;
+	r_rbtree_cont_insert (dfg->reg_vars, rv, _rv_ins_cmp, NULL);
+	return true;
 }
 
+
+static int _rv_find_cmp (void *incoming, void *in, void *user) {
+	EsilDFGRegVar *rv_incoming = (EsilDFGRegVar *)incoming;
+	EsilDFGRegVar *rv_in = (EsilDFGRegVar *)in;
+
+// first handle the simple cases without intersection
+	if (rv_incoming->to < rv_in->from) {
+		return -1;
+	}
+	if (rv_in->to < rv_incoming->from) {
+		return 1;
+	}
+
+/*
+     =incoming=
+=========in=========
+*/
+	if (rv_in->from <= rv_incoming->from && rv_incoming->to <= rv_in->to) {
+		return 0;
+	}
+
+/*
+   =incoming=
+      =in=
+
+enqueue the non-intersecting ends in the todo-queue
+*/
+	RQueue *todo = (RQueue *)user;
+	if (rv_incoming->from < rv_in->from && rv_in->to < rv_incoming->to) {
+		// lower part
+		EsilDFGRegVar *rv = R_NEW (EsilDFGRegVar);
+		rv[0] = rv_incoming[0];
+		rv->to = rv_in->from - 1;
+		r_queue_enqueue (todo, rv);
+		// upper part
+		rv = R_NEW (EsilDFGRegVar);
```

Comment:
nullchk

Reply:
cannot do null checks in this callback rbtree does not allow aborting operation